bokeh/document.py

```diff
@@ -531,6 +532,100 @@ def _initialize_references_json(cls, references_json, references):
 
             instance.update_from_json(obj_attrs, models=references)
 
+    @classmethod
+    def _event_for_attribute_change(cls, changed_obj, key, new_value, value_refs, doc):
+      event = dict(
+        kind='ModelChanged',
+        model=dict(id=changed_obj['id'], type=changed_obj['type']),
+        attr=key,
+        new=new_value,
+      )
+      HasProps._json_record_references(doc, new_value, value_refs, True) # True = recurse
+      return event
+
+    @classmethod
+    def _events_to_sync_objects(cls, from_obj, to_obj, value_refs, doc):
+        from_keys = set(from_obj['attributes'].keys())
+        to_keys = set(to_obj['attributes'].keys())
+        removed = from_keys - to_keys
+        added = to_keys - from_keys
+        shared = from_keys & to_keys
+
+        events = []
+        for key in removed:
+          # we don't really have a "remove" event - not sure this ever happens even -
+          # but treat it as equivalent to setting to null
+          #events.append(Document._event_for_attribute_change(from_obj, key, None, value_refs))
+          raise RuntimeError("'RemoveAttribute' events not supported")
```

Comment:
delete the comment?

It looks like this is in fact an assertion failure (i.e. by using the API it should be impossible to trigger) because `Property.__delete__` removes the application-set value but it doesn't remove the attribute itself since the default remains.
So we could say `raise RuntimeError("bug: should not be possible to delete attribute %s" % key)` or something